DataFormats/Provenance/src/IndexIntoFile.cc

```diff
@@ -1959,19 +1998,30 @@ namespace edm {
   }
 
   bool IndexIntoFile::IndexIntoFileItrEntryOrder::lumiEntryValid(int index) const {
+    assert(index >= 0 && index < indexedSize_);
     auto entry = runOrLumisEntry(index).entry();
     if (entry == invalidEntry) {
-      if (index + 1 < size()) {
+      // Practically, this function serves its intended purpose, but the behavior is
+      // inconsistent with the name. When we try to initialize indexToLumi_ in
+      // initializeLumi and hit the last lumi without finding any valid ones this
+      // function calls the last one valid even if its entry is invalid.
+      // This is because something needs to be processed for the lumi.
+      // In this case, shouldWeProcessRunOrLumi will return false to let
+      // the code using IndexIntoFile know to not actually try to process
+      // the lumi.
+      if (index + 1 < indexedSize_) {
         if (runOrLumisEntry(index).lumi() != runOrLumisEntry(index + 1).lumi()) {
           return true;
         }
+      } else if (index + 1 == indexedSize_) {
```

Comment:
This is used by initializeLuminosityBlock which is asking if the lumi entry at "index" is a good place to start the iteration over lumi entries at beginLumi. It calls lumiEntryValid(index) repeatedly until  it returns true, incrementing "index" by 1 in each call.

We have to use one lumi entry to force a beginLumi transition. Best is to use the first valid one, skipping over as many invalid lumi entries as possible. Sometimes there are not any valid ones in a sequence so we just take the last lumi entry. The last one is identified by looking and seeing that the next one is a different lumi or a run or in this case at the end of the container.

Execution might actually never get here because I think the last sequence of lumi entries will always have a valid lumi entry before the end of the container. I am being paranoid and covering all the possible cases, but I don't think we'll ever hit the end of the container. I could make this an assert...

Reply:
We could rename the function, maybe lumiIterationStartingPoint?